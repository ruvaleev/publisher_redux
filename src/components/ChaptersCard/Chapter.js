import React from 'react';
import { connect } from 'react-redux';

import Title from '../shared/Title';
import SubsectionsList from '../SubsectionsCard';
import isReady from '../shared/isReady';
import { addSubsection } from '../../redux/slices/books';

function SubsectionForm ({onSubmit, title, className}) {
  return (
    <td className={className}>
      <form onSubmit={onSubmit}>
        <input type='text' name='title' className='mr-4' data-testid='Add subsection'/>
        <button>{title}</button>
      </form>
    </td>
  )
}

const Chapter = ({chapter, editable, addSubsection, bookId}) => {
  return (
    <tr key={chapter.id} className={`flex flex-col ${isReady(chapter) && 'text-green-500'}`}>
      <Title title={chapter.title} className='flex'/>
      <SubsectionsList bookId={bookId} chapterId={chapter.id}/>
      {
        editable &&
          <SubsectionForm
            onSubmit={(e) => {
              e.preventDefault();
              addSubsection({bookId: bookId, chapterId: chapter.id, title: e.target.title.value });
              e.target.title.value = '';
            }}
            title='Add subsectiona'
            className='pl-4'/>
      }
    </tr>
  )
}

const mapStateToProps = (state) => ({
  editable: state.tableOfContentsReducer.editable
})

const mapDispatchToProps = (dispatch) => ({
  addSubsection: (data) => dispatch(addSubsection(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(Chapter);
