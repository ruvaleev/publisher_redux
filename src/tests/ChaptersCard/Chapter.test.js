import React from 'react';
import { render } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Chapter from '../../components/ChaptersCard/Chapter';
import Store from '../shared/Store';

function renderWithStore(store, chapter, bookId) {
  return render(
    <Provider store={store}>
      <table>
        <tbody>
          <Chapter chapter={chapter} bookId={bookId}/>
        </tbody>
      </table>
    </Provider>
  );
}

describe('ChaptersCard', () => {
  const mockStore = configureStore([]);
  let store
  let chapter
  let bookId
  let component

  describe('Chapter', () => {
    it("doesn't render subsection form if state editable is false", () => {
      store = mockStore(Store({ tableOfContentsReducer: { editable: false } }));
      chapter = store.getState().booksReducer.present.books[0].chapters[0]
      bookId = store.getState().booksReducer.present.books[0]._id
      component = renderWithStore(store, chapter, bookId)
   
      expect(component.queryByText('Add subsection')).not.toBeInTheDocument()
    });

    it('renders subsection form if state editable is true', () => {
      store = mockStore(Store({ tableOfContentsReducer: { editable: true } }));
      chapter = store.getState().booksReducer.present.books[0].chapters[0]
      bookId = store.getState().booksReducer.present.books[0]._id
      component = renderWithStore(store, chapter, bookId)
   
      expect(component.queryByText('Add subsection')).toBeInTheDocument()
    });
  });
})
