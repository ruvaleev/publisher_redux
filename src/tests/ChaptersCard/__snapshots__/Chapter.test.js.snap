// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChaptersCard Chapter renders correctly Chapter 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <tbody>
          <tr
            class="flex flex-col false"
          >
            <td
              class="flex"
            >
              глава 1
            </td>
            <td>
              <table>
                <tbody>
                  <tr>
                    <td
                      class="pl-4"
                    >
                      подраздел 1
                    </td>
                    <td>
                      <input
                        type="checkbox"
                      />
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
            <td
              class="pl-4"
            >
              <form>
                <input
                  class="mr-4"
                  data-testid="Add subsection"
                  name="title"
                  type="text"
                />
                <button>
                  Add subsectiona
                </button>
              </form>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <tbody>
        <tr
          class="flex flex-col false"
        >
          <td
            class="flex"
          >
            глава 1
          </td>
          <td>
            <table>
              <tbody>
                <tr>
                  <td
                    class="pl-4"
                  >
                    подраздел 1
                  </td>
                  <td>
                    <input
                      type="checkbox"
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
          <td
            class="pl-4"
          >
            <form>
              <input
                class="mr-4"
                data-testid="Add subsection"
                name="title"
                type="text"
              />
              <button>
                Add subsectiona
              </button>
            </form>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
